### ISSUE TITLE
better error message?

### ISSUE THREAD

**Issue by sanette** [ocaml/ocaml#11543]
Hi,
recently there was another discussion in the forum about how to improve error messages, and shortly after indeed I ran into this one

 ```
# ListLabels.map ~f:(fun i -> i+1). [1;2;3];;
Line 1, characters 18-32:
1 | ListLabels.map ~f:(fun i -> i+1). [1;2;3];;
                      ^^^^^^^^^^^^^^
Error: This expression should not be a function, the expected type is string
```

(the error is the dot after the parenthesis)

The first question is: why?  I'm sure there is a logical explanation. (the dot is used to access chars inside the string like a.(i), but why is this picked up here? there is not even a parenthesis after the dot)

But the most important is: can we improve this ? It is clear to any "human" that we're not dealing with strings here...

**Comment by Octachron** [comment]
The error is precisely that you are trying to apply the string indexing operator `.[]` to `(fun i -> i +1)` which is not a string.
A code formatter tool might be a better to way to spot  that `~f:(fun i -> i + 1). [ 1;2;3]` reads as `~f: ( (fun i -> i + 1).[(1;2;3)])`?

**Comment by sanette** [comment]
I see. Thanks! I didn't realise that `"abcde".[1;2;3]` was a valid statement, my bad. 

But still, are you really satisfied with the current error message?

An error message is supposed to tell what should be corrected. Here the suggestion is that this expression "should be a string",
but of course `ListLabels.map ~f:"abcdef". [1;2;3];;` is not valid! Could the compiler be smart enough to detect this before choosing the error message? [I know this is a very naive remark, I know nothing about the compiler internals, but sometimes naive remarks can be interesting...]

As you said, a code formatter could be helpful if it could show how the code is parsed. My current emacs+merlin config does not do this. Or, couldn't the compiler itself rewrite the code as you said when emitting the error?

AFAIC, it would be less confusing to simply say "syntax error". This is what happens when you type:
```
# (fun i -> i+1). 1;;

Line 1, characters 16-17:
1 | (fun i -> i+1). 1;;
                    ^
Error: Syntax error
```
this is one is clear [better that saying "1 is not a valid index for a string" hehe  ;)]
although I understand the logic that makes the compiler emit the non-clear message.

**Comment by sanette** [comment]
Ah, a very simple suggestion that could "solve" the issue: could the error message be changed to something like
```
"Error: Because of the dot (.) operator, this expression should not be a function, the expected type is string"
```
?

**Comment by gasche** [comment]
I wonder whether we could detect when operators that are morally a single lexeme (here `.[`) have visible space inserted in the middle, and warn (or even possibly fail?) in this case. They are full of corner cases that introduce weird ambiguities.

**Comment by sanette** [comment]
Speaking of the dot ".", some other confusing messages:

Suppose you make a typo when writing this:
```
# Float.add 180. 2.;;
- : float = 182.
```
This one is ok:
```
# Float.add 180 . 2.;;
Line 1, characters 16-18:
1 | Float.add 180 . 2.;;
                    ^^
Error: Syntax error
```
This one is confusing and illogical (?). A constant like 180 cannot have a record field... (right?)
```
# let x = 2.;;
val x : float = 2.
# Float.add 180 . x;;
Line 1, characters 16-17:
1 | Float.add 180 . x;;
                    ^
Error: Unbound record field x
```

This one I cannot say, I don't even know what a modifier is ;)
```
# Float.add 180.x;;
Line 1, characters 10-15:
1 | Float.add 180.x;;
              ^^^^^
Error: Unknown modifier 'x' for literal 180.x
```

This variation is even more confusing. (it occured to me in "real life" and it took me a while to see what was happening...)
Again it's not very logical, the compiler could make a difference between a "variable, potentially of type int" and a literal constant "180" which cannot have any record fields.
```
# type t = { x : float };;
type t = { x : float; }

(* some code to make you forget this... and then *)

# let x = 2.;;
val x : float = 2.
# Float.add 180 . x;;
Line 1, characters 10-13:
1 | Float.add 180 . x;;
              ^^^
Error: This expression has type int but an expression was expected of type t
# 
```

**Comment by Octachron** [comment]
Type error messages report conflicts between the types of various part of your program. When you are trying to apply the string indexing operator to a function, there is a clear contradiction that the typechecker can report. But it is impossible to know which side of the contradiction is erroneous, because such distinction only exists in the mind of the programmer. Going backward and trying to fix the user program to end up with a well-typed program (if that was possible) would open even more avenues for opaque  error message.

I have the impression from your later examples that you are mostly confused by the order of typing of record projection.
When you write `e . label`, the typechecker will check the record projection first without looking at the expression `e`.
Thus in your example `1.x` , the first contradiction that the typechecker sees is that there are no `x` record field. Similarly, when a field `x` is in scope, `1.x` is trying to apply the `.x` projection to the integer `1`. There is an obvious contradiction, but there are no way to know which side is the right one in the mind of the compiler.

Maybe some actionable points could be to try to follow a little more closely the source order when typechecking? I have the impression that the fact that infix operators are typed first during application is one real source of confusion.

Similarly, maybe the error message wording could be improved to emphasize the fact that there is a contradiction between two types, and it is not clear which side is correct (or in fact, if there are any correct side)?

**Comment by xavierleroy** [comment]
One other thing that might have confused the original poster is that `a.  [b]` (with one or more spaces between `.` and `[`) reads like `a.[b]`.  There may be reasons why we're not lexing `.[` as a single token, but I can't remember them right now.

**Comment by sanette** [comment]
> that `a. [b]` (with one or more spaces between `.` and `[`) reads like `a.[b]`.

exactly. also for record fields: what is the reason for allowing a space after the dot? One could declare that  `z. re` is a syntax error, that would simplify error checking.

**Comment by Octachron** [comment]
OCaml syntax is white space insensitive: white spaces are always allowed between tokens. From that point of view, the expression `x . A.Very.Long.Path.field` is not different from `x = A.Very.Long.Path.value`.

**Comment by sanette** [comment]
I understand, but my question was: why? what was the motivation for this rule? (I mean, why should the rule for "=" be the same as the rule for "." ?)

**Comment by Octachron** [comment]
Because they are both infix operators? Why should they follow different rules?

**Comment by sanette** [comment]
> Why should they follow different rules?

maybe because they are used for quite different things? maybe because they introduce difficulties for generating helpful error message?

**Comment by sanette** [comment]
Going back to the `.[` token and Xavier's question, do you know the reason for allowing a space in between?

**Comment by gasche** [comment]
Thanks to the annoying bot I reviewed this issue:

- #12160 was proposed to solve this issue but the implementation approach is not good enough, and I think we will end up closing it without merging.
- I remain in favor of forbidding spaces (or comments) in the middle of operators, to avoid this whole class of issues. This is a more invasive change with a potential backward-compatibility impact.
